File: ios/LoveWidgets/WidgetModule/WidgetModule.m
//
//  WidgetModule.m
//  LoveWidgets Native Module Bridge
//
//  Objective-C bridge for WidgetModule Swift class
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(WidgetModule, NSObject)

RCT_EXTERN_METHOD(reloadWidget:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(reloadTimelines:(NSString *)kind
                  resolve:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(saveWidgetData:(NSString *)key
                  value:(NSString *)value
                  resolve:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: widgets/ios/WidgetModule.swift
//
//  WidgetModule.swift
//  LoveWidgets Native Module
//
//  Native module to reload iOS widgets from React Native
//

import Foundation
import WidgetKit
import React

@objc(WidgetModule)
class WidgetModule: NSObject, RCTBridgeModule {
  
  static func moduleName() -> String! {
    return "WidgetModule"
  }
  
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  @objc
  func reloadWidget(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
        resolve(true)
      } else {
        reject("UNSUPPORTED", "Widgets require iOS 14.0 or later", nil)
      }
    }
  }
  
  @objc
  func reloadTimelines(_ kind: String, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadTimelines(ofKind: kind)
        resolve(true)
      } else {
        reject("UNSUPPORTED", "Widgets require iOS 14.0 or later", nil)
      }
    }
  }
}

